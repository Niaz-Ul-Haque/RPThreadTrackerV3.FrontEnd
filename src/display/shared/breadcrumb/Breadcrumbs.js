// #region imports
import React from 'react';
import { Route } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Breadcrumb } from 'reactstrap';
import BreadcrumbsItem from './BreadcrumbsItem';
// #endregion imports

const propTypes = {
	location: PropTypes.shape({
		location: PropTypes.shape({ pathname: PropTypes.string })
	}).isRequired
};

class Breadcrumbs extends React.Component {
	getPaths(pathname) {
		const paths = ['/'];
		if (pathname === '/') {
			return paths;
		}
		pathname.split('/').reduce((prev, curr) => {
			const currPath = `${prev}/${curr}`;
			paths.push(currPath);
			return currPath;
		});
		return paths;
	}
	render() {
		const { location: { pathname } } = this.props;
		const paths = this.getPaths(pathname);
		// eslint-disable-next-line react/no-array-index-key
		const items = paths.map((path, i) => <Route key={i} path={path} component={BreadcrumbsItem} />);
		return (<Breadcrumb>{items}</Breadcrumb>);
	}
}

Breadcrumbs.propTypes = propTypes;
export default Breadcrumbs;
