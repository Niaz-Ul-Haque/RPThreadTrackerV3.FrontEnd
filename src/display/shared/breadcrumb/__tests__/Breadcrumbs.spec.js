// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import Breadcrumbs from '../Breadcrumbs';
// #endregion imports

jest.mock('../../styled/Breadcrumb', () => 'Breadcrumb');
jest.mock('../BreadcrumbsItem', () => () => 'BreadcrumbsItem');

const createTestProps = propOverrides => ({
	location: { pathname: '' },
	...propOverrides
});
describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot with empty string', () => {
			const props = createTestProps();
			const jsx = (<Breadcrumbs {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with root path', () => {
			const props = createTestProps({ location: { pathname: '/' } });
			const jsx = (<Breadcrumbs {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with one segment path', () => {
			const props = createTestProps({ location: { pathname: '/test' } });
			const jsx = (<Breadcrumbs {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with more than one segment path', () => {
			const props = createTestProps({ location: { pathname: '/test/test2' } });
			const jsx = (<Breadcrumbs {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
	});
	describe('children', () => {
		it('should be rendered for each route segment', () => {
			const props = createTestProps({ location: { pathname: '/test/test2' } });
			const jsx = (<Breadcrumbs {...props} />);
			const element = shallow(jsx);
			const routes = element.find('Route');
			expect(routes).toHaveLength(3);
		});
	});
});

describe('behavior', () => {
	describe('getPaths', () => {
		it('should return root path for empty string', () => {
			const props = createTestProps();
			const element = shallow(<Breadcrumbs {...props} />);
			const paths = element.instance().getPaths('');
			expect(paths).toHaveLength(1);
			expect(paths).toContain('/');
		});
		it('should return root path for root path', () => {
			const props = createTestProps();
			const element = shallow(<Breadcrumbs {...props} />);
			const paths = element.instance().getPaths('/');
			expect(paths).toHaveLength(1);
			expect(paths).toContain('/');
		});
		it('should return array for one segment path', () => {
			const props = createTestProps();
			const element = shallow(<Breadcrumbs {...props} />);
			const paths = element.instance().getPaths('/test');
			expect(paths).toHaveLength(2);
			expect(paths).toContain('/');
			expect(paths).toContain('/test');
		});
		it('should return array for more than one segment path', () => {
			const props = createTestProps();
			const element = shallow(<Breadcrumbs {...props} />);
			const paths = element.instance().getPaths('/test/test2');
			expect(paths).toHaveLength(3);
			expect(paths).toContain('/');
			expect(paths).toContain('/test');
			expect(paths).toContain('/test/test2');
		});
	});
});
