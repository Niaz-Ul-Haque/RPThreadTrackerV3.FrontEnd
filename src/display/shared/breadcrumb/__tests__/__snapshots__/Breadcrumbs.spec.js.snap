// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot with empty string 1`] = `
<Styled(Breadcrumb)>
  <Route
    component={[Function]}
    key="/"
    path="/"
  />
</Styled(Breadcrumb)>
`;

exports[`rendering snapshots should render valid snapshot with more than one segment path 1`] = `
<Styled(Breadcrumb)>
  <Route
    component={[Function]}
    key="/"
    path="/"
  />
  <Route
    component={[Function]}
    key="/test"
    path="/test"
  />
  <Route
    component={[Function]}
    key="/test/test2"
    path="/test/test2"
  />
</Styled(Breadcrumb)>
`;

exports[`rendering snapshots should render valid snapshot with one segment path 1`] = `
<Styled(Breadcrumb)>
  <Route
    component={[Function]}
    key="/"
    path="/"
  />
  <Route
    component={[Function]}
    key="/test"
    path="/test"
  />
</Styled(Breadcrumb)>
`;

exports[`rendering snapshots should render valid snapshot with root path 1`] = `
<Styled(Breadcrumb)>
  <Route
    component={[Function]}
    key="/"
    path="/"
  />
</Styled(Breadcrumb)>
`;
